### PR TITLE
#299 Code Optimization - replace hard-coded `webgl format` with loop

### PR DESCRIPTION
**Pull request by ProcrastinatorCp**
## Description

- Code Optimization. (Using loop to generate `webgl shader precision` )
- Added `error` log on `uglify` 

## Related Issue
https://github.com/Valve/fingerprintjs2/issues/299

## Motivation and Context

Reducing the file size of the block changed. 
- File size before : **36.2 kB** 
- File size after : **32.1 kB**

## Types of changes

- [x] Code Optimization
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [x] All new and existing tests passed.
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,5 +37,8 @@
       "specs/lib",
       "specs/phantomjs-testrunner.js"
     ]
+  },
+  "dependencies": {
+    "uglify-es": "^3.3.9"
```
Comment (ProcrastinatorCp) [inline package.json]: added dependency for es6

File: gulpfile.js [hunk 2]
```diff
@@ -24,6 +26,8 @@ gulp.task('minify', function () {
         output: {
           ascii_only: true
         }
+      }).on('error', function(e) {
```
Comment (ProcrastinatorCp) [inline gulpfile.js]: While debugging, the errors where not shown with file name and line number. Added error for easier debugging. Let me know if not needed.

File: gulpfile.js [hunk 3]
```diff
@@ -1,7 +1,9 @@
 const gulp = require('gulp')
 const standard = require('gulp-standard')
 const rename = require('gulp-rename')
-const uglify = require('gulp-uglify')
+const uglifyes = require('uglify-es')
```
Comment (ProcrastinatorCp) [inline gulpfile.js]: using `uglify-es` for ES6.

File: fingerprint2.js [hunk 4]
```diff
@@ -899,6 +871,31 @@
         return null
       }
     },
+    getWebGlShaderPrecisionFormat: function () {
+      try {
+        var gl = this.getWebglCanvas()
+        if (!gl) { return null }
+        var resultArray_ = []
+        for (const numType of ['FLOAT', 'INT']) {
```
Comment (jonashaag) [inline fingerprint2.js]: Don't use `for ... of` loops, it's not supported by all browsers. My suggestion is to keep the unfolded, hard-coded list of combinations, i.e.

```
[
['FLOAT', 'VERTEX', 'HIGH', 'precision'],
['FLOAT', 'VERTEX', 'HIGH', 'rangeMin'],
...
['FRAGMENT', 'LOW', 'rangeMax']
]
```

Or use `.forEach`, which is polyfilled for browsers that don't support it.

File: gulpfile.js (hunk 3 above)
Comment (jonashaag) [reply gulpfile.js]: Remove this; we're not using ES6 yet.

File: package.json (hunk 1 above)
Comment (jonashaag) [reply package.json]: see above

**Comment by jonashaag** [comment]
Please don't recompile .min.js

**Comment by ProcrastinatorCp** [comment]
Hi @jonashaag 
Thanks for reviewing the PR. I have fixed as per the feedback. 
> Please don't recompile .min.js
I couldn't understand the above. Would you please explain.

File: fingerprint2.js [hunk 5]
```diff
@@ -899,6 +870,34 @@
         return null
       }
     },
+    getWebGlShaderPrecisionFormat: function () {
```
Comment (jonashaag) [inline fingerprint2.js]: inline this function into the place where it's called, and get rid of the array copying. Directly push to the result array.

File: specs/specs.js [hunk 6]
```diff
@@ -261,6 +261,44 @@ describe('Fingerprint2', function () {
       })
     })
 
+    describe('webgl shader precision format', function () {
+      it('checks webgl shader precision format loop', function (done) {
+
```
Comment (ProcrastinatorCp) [inline specs/specs.js]: Checking if all the corresponding entries are present. 
Couldn't think of a better approach for this spec. Please suggest if any.

**Comment by ProcrastinatorCp** [comment]
> If you're interested you can also try to make the series ... 

On a first scan, the block doesn't seem to be too similar to reduce. Will give it a try in near future.

File: specs/specs.js (hunk 6 above)
Comment (jonashaag) [reply specs/specs.js]: The problem with this kind of test is that it duplicates the code logic 1:1. Tests shouldn't try to be smart, as every little bit of smartness increases the chance of a bug in the tests.

My suggestion: Copy and paste the list of entries from the old fingerprint.js file into the tests, and compare them. It's find to have a hardcoded list of entries in the tests.

Comment (ProcrastinatorCp) [reply specs/specs.js]: Cool. Done.

**Comment by jonashaag** [comment]
Merged with some minor changes, thanks!

**Comment by sirian** [comment]
@ProcrastinatorCp 
You could also refactor this part if you wish)
image

for example

```
result.push('webgl max cube map texture size:' + gl.getParameter(gl.MAX_CUBE_MAP_TEXTURE_SIZE))
      result.push('webgl max fragment uniform vectors:' + gl.getParameter(gl.MAX_FRAGMENT_UNIFORM_VECTORS))
      result.push('webgl max render buffer size:' + gl.getParameter(gl.MAX_RENDERBUFFER_SIZE))
```


```
var wql = [
['max cube map texture size:', gl.MAX_CUBE_MAP_TEXTURE_SIZE],
['max fragment uniform vectors:',gl.MAX_FRAGMENT_UNIFORM_VECTORS],
['max render buffer size:', gl.MAX_RENDERBUFFER_SIZE]
...
]

this_.each(wgl, function (pair) {
    result.push('webql ' + pair[0], gl.getParameter(pair[1]))
});

**Comment by jonashaag** [comment]
You could even infer the human readable name from the constant name

**Pull request closed without merging**